lua/entities/gmod_tardis/modules/sh_chameleon.lua

```diff
@@ -0,0 +1,363 @@
+
+        local new_int_z_offset
+
+        if id == false then
+            new_int_z_offset = 0
+        else
+            new_int_z_offset = 0.5 * extportal:GetHeight() - intp_midtofloor
+        end
+
+        intportal:SetExitPosOffset(Vector(intp_offset.x, intp_offset.y, intp_offset_real_z + new_int_z_offset))
+        self:SetData("chameleon_intportal_z_offset", new_int_z_offset)
+
+
+        -- aligning exterior portal
+
+        local prev_ext_z_offset = self:GetData("chameleon_extportal_z_offset", 0)
+
+        local extp_offset = extportal:GetExitPosOffset()
+        local extp_offset_real_z = extp_offset.z - prev_ext_z_offset
+
+        local new_ext_z_offset
+        if id == false then
+            new_ext_z_offset = 0
+        else
+            new_ext_z_offset = intp_midtofloor - 0.5 * extportal:GetHeight()
+        end
+
+        extportal:SetExitPosOffset(Vector(extp_offset.x, extp_offset.y, extp_offset_real_z + new_ext_z_offset))
+        self:SetData("chameleon_extportal_z_offset", new_ext_z_offset)
+
+        -- exterior parts replacement
+        for k,v in pairs(self:GetParts()) do
+            if IsValid(v) then
+                v:Remove()
+            end
+        end
+
+        TARDIS:SetupParts(self)
+        TARDIS:SetupRandomSkin(self)
+        self.Fallback=self.metadata.Exterior.Fallback
+
+        self:CallCommonHook("ExteriorChanged", id)
+        self:SendMessage("exterior_changed", {id})
+
+        self:SetData("chameleon_active", (id ~= false), true)
+
+        TARDIS:StatusMessage(ply, "Chameleon.Status", (id ~= false), "Chameleon.Status.Activated", "Chameleon.Status.Deactivated")
+    end)
+end
+
+function ENT:RetryChameleon(animate)
+    local id = self:GetData("chameleon_selected_exterior")
+    local ply = self:GetData("chameleon_exterior_last_selector")
+    if id ~= nil then
+        self:ChangeExterior(id, animate, ply, true)
+    end
+end
+
+ENT:AddHook("ToggleDoor", "chameleon", function(self,open)
+    if open then return end
+    self:RetryChameleon(true)
+end)
+
+ENT:AddHook("PowerToggled", "chameleon", function(self,on)
+    if not on then return end
+    self:RetryChameleon(true)
+end)
+
+ENT:AddHook("MatStart", "chameleon", function(self)
+    self:RetryChameleon(false)
+end)
+
+ENT:AddHook("CanToggleDoor", "chameleon", function(self)
+    if self:GetData("chameleon_changing") then
+        return false
+    end
+end)
+
+ENT:AddHook("ShouldDrawShadow", "chameleon", function(self)
+    if self:GetData("chameleon_changing") then
+        return false
+    end
+end)
+
+
+ENT:AddHook("CanChangeExterior", "chameleon", function(self, id, retry)
+    if self:GetData("chameleon_changing") then
+        return false,true,"Chameleon.FailReasons.AlreadyChanging",true
+    end
+
+    if not retry then
+        local selected = self:GetData("chameleon_selected_exterior")
+        local current = self:GetData("chameleon_current_exterior", nil)
+
+        if id == selected then
```

Comment:
This still allows you to reset/select to the same exterior if it's the first one after spawning the TARDIS, pressing it again works as intended it's just the first time that doesn't work properly.

Reply:
If chameleon is active at spawn, it doesn't
It it isn't active, there's no way to tell if it's the same exterior or not

Reply:
That's fine for selecting the same exterior I suppose but especially when pressing reset when chameleon has not been activated it should not let you

Reply:
That should be fixed now as well
d317f2296d75d84c887c860916cd3a3ca1251e25